### PR TITLE
Implement soft-limit memory back-pressure for inbound reads (8.3.3)

### PR DESCRIPTION
**Pull request by leynos**
## Summary
- Adds soft-limit memory back-pressure to inbound reads to pace socket reads when per-connection budgets approach configured aggregate caps. This complements existing hard-cap enforcement (8.3.2) and preserves runtime behavior when budgets are not configured. Also includes a new backpressure module, unit tests, behavioural testing scaffolding (rstest-bdd v0.5.0), and ADR/docs updates. Minor lint-oriented adjustments to interleaved push-queue tests were also made to satisfy must_use requirements.

### Core functionality
- New: src/app/frame_handling/backpressure.rs
  - Implement should_pause_inbound_reads(state, budgets) -> bool to decide if inbound reads should pause based on current buffered bytes and configured budgets.
  - Internal helper logic uses the smallest active aggregate cap (bytes_per_connection vs. bytes_in_flight) as the governing threshold and a private pause-duration constant.
- Updated: src/app/frame_handling/mod.rs
  - Export the new backpressure module.
- Updated: src/app/inbound_handler.rs
  - Before polling framed.next(), consult should_pause_inbound_reads(...) and apply a short asynchronous pause when soft pressure is active.
  - Preserve existing timeout/decode/assembly/dispatch ordering and error handling.

### Testing
- New: src/app/frame_handling/backpressure_tests.rs
  - rstest-based unit tests covering: no budgets -> no pause; below threshold -> no pause; at/above threshold -> pause; dual-budget scenario uses the smallest cap.
- Updated/added behavioural test scaffolding (rstest-bdd v0.5.0):
  - memory_budget_backpressure feature + fixtures/steps/scenarios (as described in the exec plan).
- Minor lint adjustments to satisfy must_use in interleaved push-queue tests (wrapping/boxing push calls as needed).

### Documentation
- Updated: docs/adr-002-streaming-requests-and-shared-message-assembly.md
  - Record decisions and approach for 8.3.3 soft-limit back-pressure.
- Updated: docs/users-guide.md
  - Explain soft-limit runtime behaviour for library consumers and its relation to hard-cap enforcement.
- Updated: docs/roadmap.md
  - Mark 8.3.3 as done only after all gates pass.

### Artefacts
- Added: src/app/frame_handling/backpressure.rs
- Added: src/app/frame_handling/backpressure_tests.rs
- Modified: src/app/frame_handling/mod.rs
- Modified: src/app/inbound_handler.rs
- Added/Modified tests and fixtures for behavioural tests (memory_budget_backpressure.*)

## Why this change
- Enables controlled pacing of inbound reads under high buffered-byte pressure, reducing latency spikes and avoiding immediate growth of in-flight frames.
- Preserves existing hard-cap behavior as a separate, deterministic safety net (8.3.2).
- Keeps public API surface stable and avoids introducing new dependencies.

## Testing plan
- Stage A (unit): verify backpressure decision logic in isolation.
- Stage B (integration): verify inbound loop pauses and resumes as thresholds are crossed.
- Stage C (BDD): validate end-to-end behavioural scenarios with deterministic time control.
- Stage D (docs): ensure ADR, user guide, and roadmap reflect the new semantics.
- Stage E (quality gates): run full test suite, lint, and formatting checks as per repository guidelines.

## Validation and acceptance criteria
- Soft-limit reads are paused when buffered bytes reach the soft threshold and resume otherwise.
- Hard-cap discard/rejection behavior remains unchanged (8.3.2).
- Unit tests cover policy decisions; behavioural tests validate delay and resume under simulated time.
- Design docs and user guide reflect the new runtime behavior.
- Roadmap updated to show 8.3.3 completed after gates pass.

## Interfaces and dependencies
- No new public API surfaces; internal API remains stable.
- Internal module: src/app/frame_handling/backpressure.rs provides should_pause_inbound_reads.
- inbound_handler consults the helper before framed.next() and applies an async pause when soft pressure is active.

## Migration and rollback
- Edits are additive and isolated; revertable by reverting added files if needed.

## How to review
- Focus on the backpressure decision logic and its integration into the inbound loop.
- Verify tests cover edge cases and that no regression to hard-cap behavior occurs.
- Check docs for consistency with runtime behavior and gating criteria.

---
Notes: The updated plan aligns with the ADR updates and roadmap commitment for 8.3.3, and integrates soft-limit pacing into the inbound path without altering hard-cap semantics.

📎 **Task**: https://www.devboxer.com/task/7980c70a-2dd6-4450-980e-340b69fa8fd1

### CODE REVIEW

File: src/app/frame_handling/backpressure_tests.rs [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+        is_last: false,
+    };
+    let input = match FirstFrameInput::new(&header, EnvelopeRouting::default(), vec![], &body) {
+        Ok(input) => input,
+        Err(error) => panic!("test first frame input should be valid: {error}"),
+    };
+    let result = match state.accept_first_frame(input) {
+        Ok(result) => result,
+        Err(error) => panic!("first frame should be accepted: {error}"),
+    };
```
[the remark this answers is not in the document]
Comment (leynos) [reply src/app/frame_handling/backpressure_tests.rs]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.
